[web] Store multiple report directories with one command

csordasmarton opened the pull request:
* Allow the user to store multiple report directories with one `CodeChecker store` command.
* New metadata.json format.
* Tests for parsing and merging metadata.json files.

bruntib commented on analyzer/codechecker_analyzer/cmd/analyze.py:
```diff
@@ -607,27 +607,44 @@ def main(args):
+            'output_path': args.output_path,
+            'result_source_files': {},
+            'analyzers': {}
+        }]}
```
I believe the `tools` will be UBSanitizer, cppcheck and other analyzers, correct? Will the format of a tools' info be flexible? I mean for example `analyzers` attribute makes sense only in CodeChecker where we have ClangSA and ClangTidy. Will cppcheck have an `analyzers` attribute with exactly one element? Or will it be possible to eliminate this unnecessary level at other tools? Or will CodeChecker handle this logic for all tools?

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- web/server/tests/unit/test_metadata_merge.py  4
- web/server/codechecker_server/metadata.py  10
- analyzer/codechecker_analyzer/cmd/analyze.py  3
- web/server/tests/unit/test_metadata_parser.py  3
- web/client/codechecker_client/metadata.py  6
         

Clones added
============
- web/server/tests/unit/test_metadata_merge.py  4
- web/server/codechecker_server/metadata.py  1
- web/server/tests/unit/test_metadata_parser.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/codechecker/codechecker/pullRequest?prid=4375788&bid=14850412)

gyorb commented on analyzer/codechecker_analyzer/analyzer.py:
```diff
@@ -131,7 +129,7 @@ def __get_statistics_data(args, manager):
 
 
-def perform_analysis(args, skip_handler, context, actions, metadata):
+def perform_analysis(args, skip_handler, context, actions, metadata_tool):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [perform_analysis is too complex (22)](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=4375788)

gyorb commented on analyzer/codechecker_analyzer/analyzer.py:
```diff
@@ -63,7 +63,7 @@ def create_actions_map(actions, manager):
 
 
-def __get_analyzer_version(context, analyzer_config_map):
+def __get_analyzer_version(context, analyzer_cfg):
```
Could we reduce the number of places we get the version of an analyzer? There is already a [version.py](https://github.com/Ericsson/codechecker/blob/4aac198f547dadc0030f21fbd874c7a76bd9b13b/analyzer/codechecker_analyzer/analyzers/clangsa/version.py) for clangsa.
Maybe we could introduce the same for clang-tidy. PR #2357 has the similar issue.

gyorb commented on analyzer/codechecker_analyzer/cmd/parse.py:
```diff
@@ -459,10 +459,21 @@ def parse(plist_file, metadata_dict, rh, file_report_map):
+        result_source_files = metadata_dict['result_source_files']
+    else:
+        for tool in metadata_dict.get('tools', {}):
+            if tool.get('name') != 'codechecker':
```
Please add  a comment why we handle 'codechecker' differently here.

gyorb commented on web/client/codechecker_client/cmd/store.py:
```diff
@@ -181,19 +182,29 @@ def __get_run_name(input_list):
-                names.append(metajson['name'])
+            if 'version' in metajson and metajson['version'] >= 2:
+                for tool in metajson.get('tools', {}):
+                    if tool.get('name') != 'codechecker':
```
Some similar comment here, why is 'codechecker' handled differently.

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- web/server/tests/unit/test_metadata_parser.py  3
- web/server/codechecker_server/metadata.py  10
- analyzer/codechecker_analyzer/cmd/analyze.py  3
- web/client/codechecker_client/metadata.py  6
- web/server/tests/unit/test_metadata_merge.py  4
         

Clones added
============
- web/server/tests/unit/test_metadata_parser.py  4
- web/server/codechecker_server/metadata.py  1
- web/server/tests/unit/test_metadata_merge.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)

gamesh411 commented on analyzer/codechecker_analyzer/cmd/analyze.py:
```diff
@@ -706,25 +706,40 @@ def main(args):
+                result_src_files = tool.get('result_source_files', {})
+                metadata_tool['result_source_files'] = \
+                    dict(metadata_tool['result_source_files'].items() +
+                         result_src_files.items())
```
Thinking forward to Python3, check this out: https://stackoverflow.com/questions/38987/how-do-i-merge-two-dictionaries-in-a-single-expression#answer-26853961
Especially the `Critiques of Other Answers` part. A merge_dicts could be implemented here on in a util package.

gyorb replied:
`metadata_tool['result_source_files'].update(result_src_files.items())` wouldn't be enough here?

gamesh411 commented on analyzer/codechecker_analyzer/cmd/parse.py:
```diff
@@ -467,10 +467,18 @@ def parse(plist_file, metadata_dict, rh, file_report_map):
+        for tool in metadata_dict.get('tools', {}):
+            result_src_files = tool.get('result_source_files', {})
+            result_source_files = dict(result_source_files.items() +
+                                       result_src_files.items())
```
Same as above.

gyorb replied:
`result_source_files.update(result_src_files.items())` wouldn't be enough here, if we just want to merge the two dictionaries?

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- web/client/codechecker_client/metadata.py  6
- web/server/tests/unit/test_metadata_merge.py  4
- web/server/codechecker_server/metadata.py  10
- analyzer/codechecker_analyzer/cmd/analyze.py  3
- web/server/tests/unit/test_metadata_parser.py  3
         

Clones added
============
- web/server/tests/unit/test_metadata_merge.py  4
- web/server/codechecker_server/metadata.py  1
- web/server/tests/unit/test_metadata_parser.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)

csordasmarton commented:
@gyorb Yes, you are right. I fixed this problem by returning an empty checker set if multiple report directories are stored in one command so detection statuses will be calculated properly. I also introduced a new functional test.

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/server/tests/unit/test_metadata_parser.py  3
- web/client/codechecker_client/metadata.py  6
- web/server/codechecker_server/metadata.py  11
         

Clones added
============
- web/server/tests/unit/test_metadata_parser.py  4
- web/tests/functional/detection_status/test_detection_status.py  4
- web/server/codechecker_server/metadata.py  1
- web/server/tests/unit/test_metadata_merge.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)

gyorb commented:
Please check the pylint errors

csordasmarton commented:
@gyorb Fixed.

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/client/codechecker_client/metadata.py  6
- web/server/tests/unit/test_metadata_parser.py  3
- web/server/codechecker_server/metadata.py  11
         

Clones added
============
- web/server/tests/unit/test_metadata_merge.py  4
- web/server/tests/unit/test_metadata_parser.py  4
- web/tests/functional/detection_status/test_detection_status.py  4
- web/server/codechecker_server/metadata.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/server/codechecker_server/metadata.py  11
- web/server/tests/unit/test_metadata_parser.py  3
- web/client/codechecker_client/metadata.py  6
- web/server/tests/unit/test_metadata_merge.py  4
         

Clones added
============
- web/server/codechecker_server/metadata.py  1
- web/server/tests/unit/test_metadata_parser.py  4
- web/server/tests/unit/test_metadata_merge.py  4
- web/tests/functional/detection_status/test_detection_status.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/server/tests/unit/test_metadata_parser.py  3
- web/server/codechecker_server/metadata.py  11
- web/client/codechecker_client/metadata.py  6
- web/server/tests/unit/test_metadata_merge.py  4
         

Clones added
============
- web/server/tests/unit/test_metadata_parser.py  4
- web/tests/functional/detection_status/test_detection_status.py  5
- web/server/codechecker_server/metadata.py  1
- web/server/tests/unit/test_metadata_merge.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)

csordasmarton commented:
@gyorb This problem does not related to this pull request. I created a patch which will solve this problem: #2639.
There were two main problems:
- The first problem is that we called the `replace_report_hash` function when the plist file does not contain any diagnostics:
https://github.com/Ericsson/codechecker/blob/a909a93049acb300c60f47002a8ce69b41ea3936/web/client/codechecker_client/cmd/store.py#L234-L241
- The second problem was that when we called this function on a plist file which does not contain any diagnostic. First we read the plist file, then truncated it and we wrote back the content if the plist file contained a diagnostic:
https://github.com/Ericsson/codechecker/blob/a909a93049acb300c60f47002a8ce69b41ea3936/tools/codechecker_report_hash/codechecker_report_hash/hash.py#L300-L312

csordasmarton commented:
@gyorb I have rebased my branch, so now everything should work fine.

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/server/tests/unit/test_metadata_parser.py  3
- web/client/codechecker_client/metadata.py  6
- web/server/codechecker_server/metadata.py  11
- web/server/tests/unit/test_metadata_merge.py  4
         

Clones added
============
- web/server/tests/unit/test_metadata_parser.py  4
- web/server/codechecker_server/metadata.py  1
- web/tests/functional/detection_status/test_detection_status.py  5
- web/server/tests/unit/test_metadata_merge.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)

gyorb commented on analyzer/codechecker_analyzer/analyzers/config_handler.py:
```diff
@@ -73,6 +75,23 @@ def analyzer_plugins(self):
+        """ Get analyzer version information. """
+        version = [self.analyzer_binary, ' --version']
+        try:
+            output = subprocess.check_output(shlex.split(' '.join(version)),
```
```suggestion
            output = subprocess.check_output(version),
```
Do we need the shlex split here? `version` is already a list of the parts of the command it should be enough.

gyorb commented on web/server/tests/unit/test_metadata_parser.py:
```diff
@@ -0,0 +1,213 @@
+    'check_commands': [
+        'CodeChecker.py analyze -o /path/to/reports /path/to/build.json'
+    ],
+    'timestamps': [1.0],
```
Shouldn't the `timestamps` be a dict with `begin` and `end` keys?

csordasmarton replied:
No. Here we will test the return value of the `__get_metadata_info_v1` and `__get_metadata_info_v2` functions. In case of timestamps it will return a list where each item will be calculated in the following way: 
https://github.com/csordasmarton/codechecker/blob/8b561a529e73b47f28cea600ca8264b8125f7f62/web/server/codechecker_server/metadata.py#L35-L38

The real metada file can be found here: https://github.com/csordasmarton/codechecker/blob/8b561a529e73b47f28cea600ca8264b8125f7f62/web/server/tests/unit/metadata_test_files/v1.json#L40-L42

And the value which can be found here 1.0 = 1571297868 - 1571297867

gyorb replied:
Ok, but in that case this not a timestamp but some duration value or time difference calculated from two timestamps.
For each check command there will be a value:
```
    'check_commands': [
        'CodeChecker.py analyze -o /path/to/reports /path/to/build.json',
        'cppcheck /path/to/main.cpp'
    ],
    'timestamps': [1.0, 1.0],
```
and with the index from the `check_commands` list we can get the corresponding analysis duration from the
`timestamps` list.
If that is right the something like `analysis_duration` would be a bit better name. Only the duration is stored not the start and end dates in the database right?

gyorb commented on web/server/tests/unit/test_metadata_parser.py:
```diff
@@ -0,0 +1,213 @@
+        'cppcheck /path/to/main.cpp',
+        'cppcheck -I /path/to/include /path/to/main.cpp'
+    ],
+    'timestamps': [1.0, 1.0],
```
The same previous dict question here.

csordasmarton replied:
Same as above.

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/server/codechecker_server/metadata.py  11
- web/client/codechecker_client/metadata.py  6
- web/server/tests/unit/test_metadata_parser.py  3
- web/server/tests/unit/test_metadata_merge.py  4
         

Clones added
============
- web/server/codechecker_server/metadata.py  1
- web/server/tests/unit/test_metadata_parser.py  4
- web/tests/functional/detection_status/test_detection_status.py  5
- web/server/tests/unit/test_metadata_merge.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/client/codechecker_client/metadata.py  6
- web/server/tests/unit/test_metadata_parser.py  3
- web/server/codechecker_server/metadata.py  11
- web/server/tests/unit/test_metadata_merge.py  4
         

Clones added
============
- web/tests/functional/detection_status/test_detection_status.py  5
- web/server/tests/unit/test_metadata_parser.py  4
- web/server/codechecker_server/metadata.py  1
- web/server/tests/unit/test_metadata_merge.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)

csordasmarton commented:
@gyorb I renamed it to `analysis_duration`. Yes, we store only duration in the database:
https://github.com/Ericsson/codechecker/blob/9e33f21258f67f6426dcd30cf426e9bdeb6187b1/web/server/codechecker_server/database/run_db_model.py#L40

gyorb commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/server/tests/unit/test_metadata_parser.py  3
- web/server/codechecker_server/metadata.py  11
- web/client/codechecker_client/metadata.py  6
- web/server/tests/unit/test_metadata_merge.py  4
         

Clones added
============
- web/tests/functional/detection_status/test_detection_status.py  5
- web/server/tests/unit/test_metadata_parser.py  4
- web/server/codechecker_server/metadata.py  1
- web/server/tests/unit/test_metadata_merge.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=4375788&bid=14850412)
